### DISCUSSION TITLE
[Ally] Passing custom params from redirect to callback

### DISCUSSION THREAD

**Discussion by HevelMc** [adonisjs/core#4801] [Help]
Hi,

Using Adonis Ally, is there a way to pass params inside the redirect url to the callback ?

For example, I have two google sign-in buttons in two places in the app, for two different type of users.
I want to create the correct user type according the button used.

A way to do this would be to use a custom state with a base64 encoding of the type (and a random key to keep it secure), and retrieve it back in the callback.

I didn't find a way to specify the state in the docs. Do you have any idea?

Thanks in advance 

```ts
public redirect({ ally, params }: HttpContext) {
  // get the type from the frontend
  const type = params.type
  // make a custom state with the type?
  return ally.use('google').getRedirectUrl()
}

public async callback({ ally, response }: HttpContext) {
  // retrieve the state and get the type back
  // use it to create the correct user type
}
```

**Comment by thetutlage** [comment]
You can use cookies for storing and retrieving the state.

**Comment by AbidKhairyAK** [comment]
you can abuse the scope or redirect params.

here's the docs:

https://docs.adonisjs.com/guides/authentication/social-authentication#defining-scopes

https://docs.adonisjs.com/guides/authentication/social-authentication#defining-redirect-query-params
